File: LeetCode/java/LC2218_Maximum_Value_of_K_Coins_from_Piles/MaximumValueOfKCoinsFromPiles.java
package LC2218_Maximum_Value_of_K_Coins_from_Piles;

import java.util.Arrays;
import java.util.List;

/**
 * <a href="https://leetcode.cn/problems/maximum-value-of-k-coins-from-piles/">Maximum Value of K Coins From Piles</a>
 * 数组；动态规划；前缀和
 */
public class MaximumValueOfKCoinsFromPiles {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.maxValueOfCoins(Arrays.asList(Arrays.asList(1, 100, 3), Arrays.asList(7, 8, 9)), 2));
    }
}

class Solution {
    public int maxValueOfCoins(List<List<Integer>> piles, int k) {
        int[] f = new int[k + 1];
        Arrays.fill(f, -1);
        f[0] = 0;
        for (List<Integer> pile : piles) {
            for (int i = k; i > 0; --i) {
                int value = 0;
                for (int t = 1; t <= pile.size(); ++t) {
                    value += pile.get(t - 1);
                    if (i >= t && f[i - t] != -1) {
                        f[i] = Math.max(f[i], f[i - t] + value);
                    }
                }
            }
        }
        return f[k];
    }
}

File: LeetCode/go/LC2218_Maximum_Value_of_K_Coins_from_Piles/maximum_value_of_k_coins_from_piles.go
package maximum_value_of_k_coins_from_piles

func maxValueOfCoins(piles [][]int, k int) int {
	f := make([]int, k+1)
	for i := range f {
		f[i] = -1
	}
	f[0] = 0
	for _, pile := range piles {
		for i := k; i > 0; i-- {
			value := 0
			for t := 1; t <= len(pile); t++ {
				value += pile[t-1]
				if i >= t && f[i-t] != -1 {
					f[i] = max(f[i], f[i-t]+value)
				}
			}
		}
	}
	return f[k]
}
